File: android_apps/player/app/src/main/kotlin/com/jieoz/lanmediawall/player/net/Envelope.kt
package com.jieoz.lanmediawall.player.net

import java.util.UUID
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * Message envelope + HMAC signing/verification — protocol_spec §2, §3.
 *
 * Mirrors windows_player/envelope.py and broker/envelope.py exactly:
 *
 *   signing_string = "{v}|{type}|{msg_id}|{ts}|{from}|{to}|{canonical_json(payload)}"
 *   sig            = HMAC_SHA256(PSK_utf8, signing_string_utf8).hexdigest()
 *
 * canonical_json is [CanonicalJson.encode] (sort_keys + compact + ensure_ascii=False).
 * Pure logic, no Android dependencies — fully unit-testable on the JVM.
 */
object Envelope {
    const val PROTOCOL_VERSION = 1

    /** The placeholder PSK shipped by default (see Settings.DEFAULT_PSK). Treated
     *  as "no real key" by [hasUsableKey] so `optional` mode degrades to `sig=""`
     *  rather than emitting a signature derived from a well-known string. */
    const val PLACEHOLDER_PSK = "CHANGE_ME_32_BYTE_RANDOM_PRESHARED_KEY"


    // §3 thresholds (mirror envelope.py)
    const val FRESH_WINDOW_MS = 30_000L
    const val FIRST_CONNECT_WINDOW_MS = 120_000L
    const val DEDUP_TTL_MS = 5L * 60 * 1000

    /** Local wall-clock epoch milliseconds (the spec's `ts`). */
    fun nowMs(): Long = System.currentTimeMillis()

    fun signingString(
        v: Int, type: String, msgId: String, ts: Long,
        from: String, to: String, payload: Json,
    ): String = "$v|$type|$msgId|$ts|$from|$to|${CanonicalJson.encode(payload)}"

    /**
     * §17.2 key derivation: `device_key = HMAC_SHA256(PSK, identity).digest()`.
     *
     * Returns the **32 raw bytes** of the HMAC digest — used directly as the key
     * of the next HMAC layer (do NOT hex-encode it first, §17.5). `identity` is
     * the envelope `from` string **verbatim** — no normalization/lowercasing/
     * trimming (§17.5). Mirrors broker/envelope.py::derive_key byte-for-byte.
     */
    fun deriveKey(psk: String, identity: String): ByteArray =
        hmacSha256(psk.toByteArray(Charsets.UTF_8), identity.toByteArray(Charsets.UTF_8))

    /**
     * Resolve the raw HMAC **key bytes** for the active [keyMode] (§17.2/§17.3):
     *   - [KeyMode.DERIVED] → `device_key` derived from [identity];
     *   - [KeyMode.GLOBAL]  → the PSK's UTF-8 bytes (v1.2 behaviour).
     * [identity] is the sender's own `from` when signing, or the frame's `from`
     * when verifying.
     */
    fun signingKey(psk: String, keyMode: KeyMode, identity: String): ByteArray =
        when (keyMode) {
            KeyMode.DERIVED -> deriveKey(psk, identity)
            KeyMode.GLOBAL -> psk.toByteArray(Charsets.UTF_8)
        }

    /**
     * Sign the §3 signing string. [keyMode] picks the HMAC key per §17:
     *   - GLOBAL (default): the raw PSK — byte-identical to v1.2.
     *   - DERIVED: the sender's own device_key (`HMAC(PSK, from)`), so the
     *     identity in `from` is bound into the key.
     */
    fun sign(
        psk: String, v: Int, type: String, msgId: String, ts: Long,
        from: String, to: String, payload: Json,
        keyMode: KeyMode = KeyMode.GLOBAL,
    ): String {
        val msg = signingString(v, type, msgId, ts, from, to, payload)
        return hmacSha256Hex(signingKey(psk, keyMode, from), msg)
    }

    /**
     * Sign with an **explicit raw key** — for a dk-only end (§17.4) that holds
     * its own `device_key` and never the PSK. The caller passes its stored
     * device_key bytes; in derived mode this is byte-identical to
     * `sign(psk, …, keyMode = DERIVED)` because `device_key == HMAC(PSK, from)`.
     */
    fun signWithKey(
        key: ByteArray, v: Int, type: String, msgId: String, ts: Long,
        from: String, to: String, payload: Json,
    ): String = hmacSha256Hex(key, signingString(v, type, msgId, ts, from, to, payload))

    fun hmacSha256Hex(psk: String, message: String): String =
        hmacSha256Hex(psk.toByteArray(Charsets.UTF_8), message)

    /** HMAC-SHA256 over [message] keyed by raw [key] bytes, hex-encoded. */
    fun hmacSha256Hex(key: ByteArray, message: String): String {
        val digest = hmacSha256(key, message.toByteArray(Charsets.UTF_8))
        val sb = StringBuilder(digest.size * 2)
        for (b in digest) {
            val v = b.toInt() and 0xFF
            sb.append(HEX[v ushr 4])
            sb.append(HEX[v and 0xF])
        }
        return sb.toString()
    }

    /** Raw HMAC-SHA256 digest (32 bytes) keyed by [key] over [data]. */
    fun hmacSha256(key: ByteArray, data: ByteArray): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(key, "HmacSHA256"))
        return mac.doFinal(data)
    }

    /** Construct a fully-signed outbound envelope as a [Json.Obj]. */
    fun build(
        psk: String, type: String, from: String, to: String, payload: Json,
        msgId: String = UUID.randomUUID().toString(),
        ts: Long = nowMs(),
        keyMode: KeyMode = KeyMode.GLOBAL,
    ): Json.Obj {
        val v = PROTOCOL_VERSION
        val sig = sign(psk, v, type, msgId, ts, from, to, payload, keyMode)
        return assemble(v, type, msgId, ts, from, to, sig, payload)
    }

    /** Serialize an envelope to wire text. Non-canonical separators are fine on
     *  the wire — the receiver re-canonicalizes the payload for verification. */
    fun toWire(env: Json.Obj): String = CanonicalJson.encode(env)

    /**
     * Auth-mode-aware build (§13). The `sig` field is **always present** so the
     * §2 envelope shape never changes; only its value depends on the mode:
     *   - [AuthMode.OPEN]      → `sig=""` (no key needed, zero-config);
     *   - [AuthMode.OPTIONAL]  → sign iff [psk] is a real key, else `sig=""`;
     *   - [AuthMode.REQUIRED]  → always sign.
     *
     * [hasUsableKey] decides "is there a real PSK" — an empty or the shipped
     * placeholder key counts as absent so an unconfigured device under
     * `optional` sends `sig=""` rather than a signature nobody can verify.
     */
    fun buildWithMode(
        psk: String, authMode: AuthMode, type: String, from: String, to: String,
        payload: Json,
        msgId: String = UUID.randomUUID().toString(),
        ts: Long = nowMs(),
        keyMode: KeyMode = KeyMode.GLOBAL,
    ): Json.Obj {
        val v = PROTOCOL_VERSION
        val sig = when (authMode) {
            AuthMode.OPEN -> ""
            AuthMode.OPTIONAL -> if (hasUsableKey(psk)) sign(psk, v, type, msgId, ts, from, to, payload, keyMode) else ""
            AuthMode.REQUIRED -> sign(psk, v, type, msgId, ts, from, to, payload, keyMode)
        }
        return assemble(v, type, msgId, ts, from, to, sig, payload)
    }

    /**
     * Build an envelope signed with an **explicit device_key** — the dk-only end
     * path (§17.4). The end holds its own `device_key` (from the §15 QR's `dk`)
     * and never the PSK, so it cannot run [buildWithMode]'s PSK derivation; it
     * signs directly. `open` still emits `sig=""`; `optional`/`required` sign
     * with [deviceKey] (which the caller derived once at pairing time as
     * `HMAC(PSK, from)`).
     */
    fun buildWithDeviceKey(
        deviceKey: ByteArray, authMode: AuthMode, type: String, from: String, to: String,
        payload: Json,
        msgId: String = UUID.randomUUID().toString(),
        ts: Long = nowMs(),
    ): Json.Obj {
        val v = PROTOCOL_VERSION
        val sig = when (authMode) {
            AuthMode.OPEN -> ""
            AuthMode.OPTIONAL, AuthMode.REQUIRED ->
                signWithKey(deviceKey, v, type, msgId, ts, from, to, payload)
        }
        return assemble(v, type, msgId, ts, from, to, sig, payload)
    }

    private fun assemble(
        v: Int, type: String, msgId: String, ts: Long,
        from: String, to: String, sig: String, payload: Json,
    ): Json.Obj = jsonObj {
        put("v", v)
        put("type", type)
        put("msg_id", msgId)
        put("ts", ts)
        put("from", from)
        put("to", to)
        put("sig", sig)
        put("payload", payload)
    }

    /** True when [psk] is a real preshared key (non-blank and not the shipped
     *  placeholder). Used by [buildWithMode] / verify gating under `optional`. */
    fun hasUsableKey(psk: String): Boolean =
        psk.isNotBlank() && psk != PLACEHOLDER_PSK


    data class Parsed(
        val v: Int,
        val type: String,
        val msgId: String,
        val ts: Long,
        val from: String,
        val to: String,
        val sig: String,
        val payload: Json,
        val payloadObj: Json.Obj,
        /**
         * True only when this frame's signature was actually recomputed and
         * matched (i.e. `mustVerify` was true and the HMAC check passed).
         * `open` mode and empty-sig `optional` frames are accepted but arrive
         * with `authed=false`. Security-sensitive handlers (e.g. §22
         * `update_app`, which can root-install an APK) MUST require
         * `authed==true` so an unauthenticated box can never be remotely
         * reflashed over the LAN.
         */
        val authed: Boolean = false,
    )

    enum class Reason { OK, SHAPE, SIG, STALE, DUP }

    data class VerifyResult(val ok: Boolean, val reason: Reason, val parsed: Parsed?)

    /**
     * Validate an inbound envelope per §3, gated by [authMode] per §13 and keyed
     * per [keyMode] (§17):
     *   1. shape check (all fields present, payload is object),
     *   2. signature check — **mode-dependent**:
     *        - [AuthMode.REQUIRED]: recompute + constant-time compare (drop on fail),
     *        - [AuthMode.OPTIONAL]: verify only when `sig` is non-empty,
     *        - [AuthMode.OPEN]: skip (accept any/empty `sig`),
     *   3. freshness window (120s on first connect, else 30s) — **always**,
     *   4. replay dedup via [replay] (5-minute LRU) — **always**.
     *
     * §17 key selection for the recompute:
     *   - [KeyMode.GLOBAL]:  HMAC key = the PSK bytes (v1.2);
     *   - [KeyMode.DERIVED]: HMAC key = `HMAC(PSK, from)` of the frame's **own**
     *     `from`, so a frame signed with identity-A's key but stamped `from=B`
     *     recomputes against B's key and is rejected — the §17.5 leak-isolation
     *     contract. broker holds the PSK and derives per-`from` statelessly.
     *
     * [verifyKeyFor] is the **dk-only** fallback (§17.4): an end that holds only
     * its own `device_key` (never the PSK) cannot derive other identities' keys.
     * When set and the PSK is not usable, it supplies the raw verify key for a
     * given `from` (e.g. a stored broker key for `from="broker"`), or null to
     * **fail closed** — an unverifiable signed frame is dropped, never accepted.
     *
     * Default [authMode] is [AuthMode.REQUIRED] / [keyMode] is [KeyMode.GLOBAL]
     * so this stays a strict v1.2 verifier for callers that don't opt in.
     */
    fun verify(
        psk: String,
        raw: String,
        replay: ReplayCache? = null,
        firstConnect: Boolean = false,
        now: Long = nowMs(),
        authMode: AuthMode = AuthMode.REQUIRED,
        keyMode: KeyMode = KeyMode.GLOBAL,
        verifyKeyFor: ((from: String) -> ByteArray?)? = null,
    ): VerifyResult {
        val root = try {
            Json.parse(raw)
        } catch (e: Exception) {
            return VerifyResult(false, Reason.SHAPE, null)
        }
        val obj = root as? Json.Obj ?: return VerifyResult(false, Reason.SHAPE, null)
        val e = obj.entries
        val required = listOf("v", "type", "msg_id", "ts", "from", "to", "sig", "payload")
        if (required.any { it !in e }) return VerifyResult(false, Reason.SHAPE, null)
        val payload = e["payload"]
        val payloadObj = payload as? Json.Obj
            ?: return VerifyResult(false, Reason.SHAPE, null)

        val v = (e["v"]).asIntOrNull() ?: return VerifyResult(false, Reason.SHAPE, null)
        val type = (e["type"]).asString() ?: return VerifyResult(false, Reason.SHAPE, null)
        val msgId = (e["msg_id"]).asString() ?: return VerifyResult(false, Reason.SHAPE, null)
        val ts = (e["ts"]).asLongOrNull() ?: return VerifyResult(false, Reason.SHAPE, null)
        val from = (e["from"]).asString() ?: return VerifyResult(false, Reason.SHAPE, null)
        val to = (e["to"]).asString() ?: return VerifyResult(false, Reason.SHAPE, null)
        val sig = (e["sig"]).asString() ?: return VerifyResult(false, Reason.SHAPE, null)

        // §13 signature gating by auth mode. `open` skips entirely; `optional`
        // verifies only a non-empty sig; `required` always verifies.
        val mustVerify = when (authMode) {
            AuthMode.OPEN -> false
            AuthMode.OPTIONAL -> sig.isNotEmpty()
            AuthMode.REQUIRED -> true
        }
        if (mustVerify) {
            // §17 key resolution. Prefer the PSK path (broker-style: can derive
            // any `from`); fall back to the dk-only resolver; else fail closed.
            val key: ByteArray? = when (keyMode) {
                KeyMode.GLOBAL -> psk.toByteArray(Charsets.UTF_8)
                KeyMode.DERIVED ->
                    if (hasUsableKey(psk)) deriveKey(psk, from)
                    else verifyKeyFor?.invoke(from)
            }
            if (key == null) return VerifyResult(false, Reason.SIG, null)
            val expected = hmacSha256Hex(key, signingString(v, type, msgId, ts, from, to, payloadObj))
            if (!constantTimeEquals(expected, sig)) {
                return VerifyResult(false, Reason.SIG, null)
            }
        }
        // `authed` is true iff we actually recomputed + matched the HMAC above.
        // open / empty-sig-optional frames pass verification but are NOT authed.
        val authed = mustVerify

        val window = if (firstConnect) FIRST_CONNECT_WINDOW_MS else FRESH_WINDOW_MS
        if (Math.abs(now - ts) > window) {
            return VerifyResult(false, Reason.STALE, null)
        }

        if (replay != null && replay.seen(msgId, now)) {
            return VerifyResult(false, Reason.DUP, null)
        }

        val parsed = Parsed(v, type, msgId, ts, from, to, sig, payloadObj, payloadObj, authed)
        return VerifyResult(true, Reason.OK, parsed)
    }

    private fun constantTimeEquals(a: String, b: String): Boolean {
        val ab = a.toByteArray(Charsets.UTF_8)
        val bb = b.toByteArray(Charsets.UTF_8)
        if (ab.size != bb.size) return false
        var r = 0
        for (i in ab.indices) r = r or (ab[i].toInt() xor bb[i].toInt())
        return r == 0
    }

    /**
     * Decode a hex string to raw bytes, or null if it isn't valid even-length
     * hex. Used to turn a pairing QR's `dk`/`bk` (device_key hex, §17.4) back
     * into the raw HMAC key bytes.
     */
    fun hexToBytes(hex: String): ByteArray? {
        val s = hex.trim()
        if (s.isEmpty() || s.length % 2 != 0) return null
        val out = ByteArray(s.length / 2)
        var i = 0
        while (i < s.length) {
            val hi = Character.digit(s[i], 16)
            val lo = Character.digit(s[i + 1], 16)
            if (hi < 0 || lo < 0) return null
            out[i / 2] = ((hi shl 4) or lo).toByte()
            i += 2
        }
        return out
    }

    private val HEX = "0123456789abcdef".toCharArray()
}

File: windows_player/envelope.py
"""Message envelope + HMAC signing/verification (protocol_spec §2, §3, §17).

Pure logic, no I/O. Used by both the WS client (out/inbound) and tests.

Signing object (§3) — **unchanged** across v1.1/v1.2/v1.3:
    f"{v}|{type}|{msg_id}|{ts}|{from}|{to}|{canonical_json(payload)}"
where canonical_json == json.dumps(payload, sort_keys=True,
separators=(",", ":"), ensure_ascii=False).

§17 key derivation — the ONLY thing v1.3 changes is the HMAC *key*:
  - key_mode="global" (v1.2): key = PSK bytes (legacy, backward compatible).
  - key_mode="derived" (v1.3): key = device_key = HMAC_SHA256(PSK, identity),
    where `identity` is the sender's `from` field verbatim (no normalization).
    sig = HMAC_SHA256(device_key, signing_string).hexdigest().
The verifier re-derives the *sender's* device_key from `from` and compares —
so a key minted for identity-A cannot sign a frame claiming `from`=identity-B
(§17.5 leak isolation).
"""

from __future__ import annotations

import hashlib
import hmac
import json
import time
import uuid
from collections import OrderedDict
from typing import Any, Callable, Dict, Optional, Tuple

PROTOCOL_VERSION = 1

# §17.3 key_mode values. global = v1.2 (PSK direct); derived = v1.3 (per-end).
KEY_MODE_GLOBAL = "global"
KEY_MODE_DERIVED = "derived"

# §3 thresholds
FRESH_WINDOW_MS = 30_000          # normal replay window
FIRST_CONNECT_WINDOW_MS = 120_000  # relaxed window right after connect
DEDUP_TTL_MS = 5 * 60 * 1000      # msg_id remembered for 5 minutes


def now_ms() -> int:
    """Local wall-clock epoch milliseconds (the spec's `ts`)."""
    return int(time.time() * 1000)


def canonical_json(payload: Dict[str, Any]) -> str:
    """Deterministic JSON used inside the signed string (§3)."""
    return json.dumps(
        payload,
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=False,
    )


def signing_string(
    v: int, type_: str, msg_id: str, ts: int, frm: str, to: str,
    payload: Dict[str, Any],
) -> str:
    return f"{v}|{type_}|{msg_id}|{ts}|{frm}|{to}|{canonical_json(payload)}"


# --- §17 key derivation ------------------------------------------------------

def _as_key_bytes(key: Any) -> bytes:
    """Coerce a key (str PSK or raw 32-byte device_key) to the HMAC key bytes.

    A `str` is UTF-8 encoded (the v1.2 PSK path); `bytes`/`bytearray` are used
    verbatim (the §17 device_key path — raw binary, NOT hex-encoded)."""
    if isinstance(key, (bytes, bytearray)):
        return bytes(key)
    return str(key).encode("utf-8")


def derive_device_key(psk: str, identity: str) -> bytes:
    """§17.2: device_key = HMAC_SHA256(PSK, identity).digest() — 32 raw bytes.

    `identity` is the sender's `from` field **verbatim** — no normalization,
    lowercasing, or trimming (§17.5). The 32-byte binary digest is used
    directly as the next HMAC's key (never hex-encode it first)."""
    return hmac.new(psk.encode("utf-8"), identity.encode("utf-8"),
                    hashlib.sha256).digest()


def hmac_hex(key: Any, signing_str: str) -> str:
    """sig = HMAC_SHA256(key, signing_str).hexdigest(). `key` is PSK str or
    raw device_key bytes (§17.2)."""
    return hmac.new(_as_key_bytes(key), signing_str.encode("utf-8"),
                    hashlib.sha256).hexdigest()


def signing_key(psk: str, frm: str, key_mode: str = KEY_MODE_GLOBAL,
                *, device_key: Optional[bytes] = None) -> Any:
    """Resolve the HMAC key the *sender* must use for a frame from `frm`.

    global → the PSK itself (v1.2). derived → the sender's own device_key:
    an explicit `device_key` (a paired, PSK-less endpoint) is used as-is, else
    it is derived from the PSK on the fly (a coordinator that holds the PSK)."""
    if key_mode == KEY_MODE_DERIVED:
        if device_key is not None:
            return device_key
        return derive_device_key(psk, frm)
    return psk


def sign(psk: str, v: int, type_: str, msg_id: str, ts: int, frm: str,
         to: str, payload: Dict[str, Any], *,
         key_mode: str = KEY_MODE_GLOBAL,
         device_key: Optional[bytes] = None) -> str:
    """HMAC-sign a frame. Defaults to v1.2 global mode (key=PSK) so existing
    callers/tests are unchanged; pass key_mode="derived" for §17."""
    s = signing_string(v, type_, msg_id, ts, frm, to, payload)
    key = signing_key(psk, frm, key_mode, device_key=device_key)
    return hmac_hex(key, s)


def build_envelope(psk: str, type_: str, frm: str, to: str,
                   payload: Dict[str, Any], *, msg_id: Optional[str] = None,
                   ts: Optional[int] = None, sign_frame: bool = True,
                   key_mode: str = KEY_MODE_GLOBAL,
                   device_key: Optional[bytes] = None) -> Dict[str, Any]:
    """Construct an outbound envelope.

    `sign_frame` controls §13 auth adaptivity: when True (default, the
    `required`/signing case) the `sig` field carries a real HMAC; when False
    (the `open` case, or `optional` with no PSK) it is the empty string. The
    envelope shape (§2) is identical either way — only `sig` differs — so a
    parser never trips on an unsigned frame (§13).

    `key_mode`/`device_key` select the §17 signing key: global (default) uses
    the PSK verbatim (v1.2); derived signs with this end's own device_key
    (`device_key` if the end is PSK-less, else derived from the PSK)."""
    v = PROTOCOL_VERSION
    msg_id = msg_id or str(uuid.uuid4())
    ts = now_ms() if ts is None else ts
    sig = (sign(psk, v, type_, msg_id, ts, frm, to, payload,
                key_mode=key_mode, device_key=device_key)
           if sign_frame else "")
    return {
        "v": v,
        "type": type_,
        "msg_id": msg_id,
        "ts": ts,
        "from": frm,
        "to": to,
        "sig": sig,
        "payload": payload,
    }


class ReplayCache:
    """LRU TTL cache of seen msg_ids for §3 dedup. Not thread-safe; callers
    on the asyncio receive loop use it from a single task."""

    def __init__(self, ttl_ms: int = DEDUP_TTL_MS, max_entries: int = 50_000):
        self.ttl_ms = ttl_ms
        self.max_entries = max_entries
        self._seen: "OrderedDict[str, int]" = OrderedDict()

    def _evict(self, now: int) -> None:
        # drop expired from the front (oldest insert first)
        while self._seen:
            mid, exp = next(iter(self._seen.items()))
            if exp <= now:
                self._seen.popitem(last=False)
            else:
                break
        while len(self._seen) > self.max_entries:
            self._seen.popitem(last=False)

    def seen(self, msg_id: str, now: Optional[int] = None) -> bool:
        """Return True if msg_id was already seen (and not expired). Records
        it as seen as a side effect when it is new."""
        now = now_ms() if now is None else now
        self._evict(now)
        if msg_id in self._seen and self._seen[msg_id] > now:
            return True
        self._seen[msg_id] = now + self.ttl_ms
        self._seen.move_to_end(msg_id)
        return False


def verify(psk: str, env: Dict[str, Any], *,
           replay: Optional[ReplayCache] = None,
           first_connect: bool = False,
           now: Optional[int] = None,
           auth_mode: str = "required",
           key_mode: str = KEY_MODE_GLOBAL,
           key_resolver: Optional[Callable[[str], Any]] = None) -> Tuple[bool, str]:
    """Validate an inbound envelope per §3, gated by §13 `auth_mode`.

    Returns (ok, reason). reason is "" on success, else a short code:
    "shape", "sig", "stale", "dup".

    §13 controls only the signature check:
      - "required" (default): always verify `sig` — preserves v1 behavior.
      - "open":               never verify `sig` (it may be "").
      - "optional":           verify only when `sig` is non-empty.
    The ts-freshness and msg_id-dedup checks run in **all** modes — they are
    replay hygiene and need no key (§13).

    §17 key selection for the sig check:
      - key_mode="global" (default): the key is the PSK (v1.2 behavior).
      - key_mode="derived": the key is the *sender's* device_key. By default it
        is derived from the PSK and the frame's `from` (a coordinator holding
        the PSK can verify any sender statelessly, §17.2). A PSK-less endpoint
        passes `key_resolver(from) -> key_bytes|None` to map an identity to a
        pre-shared verify key (e.g. the broker key from pairing); returning
        None means "no key for this sender" → the sig check fails closed."""
    now = now_ms() if now is None else now
    required = ("v", "type", "msg_id", "ts", "from", "to", "sig", "payload")
    if not isinstance(env, dict) or any(k not in env for k in required):
        return False, "shape"
    if not isinstance(env["payload"], dict):
        return False, "shape"

    sig = str(env.get("sig", ""))
    if _verify_needed(auth_mode, sig):
        frm = env["from"]
        if key_resolver is not None:
            key = key_resolver(frm)
        elif key_mode == KEY_MODE_DERIVED:
            key = derive_device_key(psk, frm)
        else:
            key = psk
        if key is None:
            # §17: no verify key for this sender (PSK-less end, unknown `from`)
            # → fail closed rather than accept an unverifiable frame.
            return False, "sig"
        expected = hmac_hex(key, signing_string(
            env["v"], env["type"], env["msg_id"], env["ts"],
            env["from"], env["to"], env["payload"]))
        # constant-time compare
        if not hmac.compare_digest(expected, sig):
            return False, "sig"

    window = FIRST_CONNECT_WINDOW_MS if first_connect else FRESH_WINDOW_MS
    try:
        ts = int(env["ts"])
    except (TypeError, ValueError):
        return False, "shape"
    if abs(now - ts) > window:
        return False, "stale"

    if replay is not None and replay.seen(str(env["msg_id"]), now):
        return False, "dup"

    return True, ""


def _verify_needed(auth_mode: str, sig: str) -> bool:
    """§13 signature-check gate. Kept here (not importing `auth`) so envelope
    stays dependency-free; `auth.should_verify` is the canonical mirror."""
    mode = auth_mode if isinstance(auth_mode, str) else "required"
    mode = mode.strip().lower()
    if mode == "open":
        return False
    if mode == "optional":
        return bool(sig)
    return True  # required / unknown → strict (safe default)
